### ISSUE TITLE
Implement case sensitivity for the GGP KB

### ISSUE THREAD

**Issue by MihaiSurdeanu** [clulab/reach#382]
Compare HashSet(case sensitive keys) vs. HashSet(case insensitive keys)

**Comment by hickst** [comment]
Using Unix command line tools, I looked at a few of our largest KB files (PFAM, PubChem, and UniprotProteins) to compare the number of unique mixed-case KB keys (cs) against the number of unique keys that result without regard to case (ci).

Since we currently canonicalize the keys as we read them in, including lowercasing them, the (ci) measure should approximate what we currently use for processing while the (cs) measure should approximate the increased number of entries we would have if we added case sensitivity.

I am not sure how valuable these measurements are since neither measure counts:
1. how many different values are contained in the value Set for each key or,
2. how many values would be duplicated across entries if entries were "separated out" by case-sensitive keys.

```
PubChem
-------
207653 TOTAL entries
203193 sort-uniq keys (cs)
196133 sort-uniq keys (ci)
  7060 more keys with case sensitivity

PFAM
----
32590 TOTAL entries
31686 sort-uniq keys (cs)
31670 sort-uniq keys (ci)
   16 more keys with case sensitivity

Uniprot
-------
2784260 TOTAL entries
 374496 sort-uniq keys (cs)
 333303 sort-uniq keys (ci)
  41193 more keys with case sensitivity
```

**Comment by MihaiSurdeanu** [comment]
It seems this is a major issue for proteins.
We could create both case-sensitive and case-insensitive keys JUST for the proteins that appear in both forms.

**Comment by hickst** [comment]
I guess the question that this approach did not answer is are the *values* (i.e. the IDs) different for the two different keys (cs vs ci) or are there just duplicates in the file, distinguished only by the case of their (otherwise matching) keys?

It's clear that Uniprot contains both. Here's a true value distinction example:
```
Hells   Q60848  Mouse
Hells   Q60848  Mus musculus
HELLS   Q9NRZ9  Homo sapiens
HELLS   Q9NRZ9  Human
```
and here's an example where the case is not used meaningfully:
```
hem1    O14092  Fission yeast
hem1    O14092  Schizosaccharomyces pombe (strain 972 / ATCC 24843)
HEM1    O94069  Candida albicans (strain SC5314 / ATCC MYA-2876)
HEM1    O94069  Yeast
HEM1    P09950  Baker's yeast
HEM1    P09950  Saccharomyces cerevisiae (strain ATCC 204508 / S288c)
HEM1    P55160  Homo sapiens
HEM1    P55160  Human
HEM1    P78698  Candida sphaerica
HEM1    P78698  Kluyveromyces lactis (strain ATCC 8585 / CBS 2359 / DSM 70799 / NBRC 1267 / NRRL Y-1140 / WM37)
HEM1    P78698  Yeast
HEM1    Q6BX71  Debaryomyces hansenii (strain ATCC 36239 / CBS 767 / JCM 1990 / NBRC 0083 / IGC 2968)
HEM1    Q6BX71  Torulaspora hansenii
HEM1    Q6BX71  Yeast
HEM1    Q6CCW0  Candida lipolytica
HEM1    Q6CCW0  Yarrowia lipolytica (strain CLIB 122 / E 150)
HEM1    Q6CCW0  Yeast
HEM1    Q6FXE3  Candida glabrata (strain ATCC 2001 / CBS 138 / JCM 3761 / NBRC 0622 / NRRL Y-65)
HEM1    Q6FXE3  Torulopsis glabrata
HEM1    Q6FXE3  Yeast
HEM1    Q75DX7  Ashbya gossypii (strain ATCC 10895 / CBS 109.51 / FGSC 9923 / NRRL Y-1056)
HEM1    Q75DX7  Eremothecium gossypii
HEM1    Q75DX7  Yeast
hem1    Q7RVY5  Neurospora crassa (strain ATCC 24698 / 74-OR23-1A / CBS 708.71 / DSM 1257 / FGSC 987)
hem1    Q92403  Agaricus bisporus
hem1    Q92403  White button mushroom
```
Both examples show the complete key set for the illustrated key.

**Comment by MihaiSurdeanu** [comment]
(I thought we eliminate the entities longer than 5 or so words; they are
unlikely to be matched anyway.)

But this could all be detected programmatically: we should build the
different case keys only when necessary, i.e., when they exists to begin
with, and when they point to different grounding ids. This could be done in
an offline preproceesing, before we load these grounding KBs.

**Comment by johnbachman** [comment]
Jumping in, perhaps out of my element here, but...as you probably already know, human gene names are "supposed" to be in all caps when written in text, (e.g. "TP53"), whereas the mouse gene names are supposed to first letter capital only ("Tp53"). However, in practice I find that that biologists (myself included) routinely refer to human gene/protein names with only the first letter capitalized (e.g., "Ras", "Bax", etc.

See, https://en.wikipedia.org/wiki/Gene_nomenclature: "the nomenclatural systems also provide for at least human-versus-nonhuman specificity by using different capitalization, although scientists often ignore this distinction, given that it is often biologically irrelevant."

I agree that this is mainly an issue with genes/proteins and not with, say, biological processes or chemicals. From perusing the texts of entities, it seems to me that the main issue is the use of prefixes attached to gene names that are differentiated from the gene name not by a hyphen, but by camel-case: "pErk" (phospho-Erk), "mErk", (mouse Erk), "hErk" (human Erk), etc. Every now and then these compound names will, in a case-insensitive match, overlap with another gene name (e.g., "PERK" http://www.genenames.org/cgi-bin/gene_symbol_report?hgnc_id=HGNC:3255).

So if it's possible to determine if something is a gene/protein vs. a chemical or process before the canonicalization step, I would look at case-sensitive matching only for the genes/proteins.

**Comment by hickst** [comment]
We always welcome your input, especially on these issues of biology. Your synopsis sounds like what I had gathered: that capitalization would be very useful if only it was a convention that was consistently followed. The use of camel-case is noted and we will take it into consideration when we re-examine how we detect the kinds of "abbreviations".

**Comment by hickst** [comment]
The latest TODO email list appears to change this task to implement case sensitivity for the GGP KB so I will change the task title accordingly. This may involve some code refactoring (and/or new code) as we designed all KBs to use the same "canonical" key transformation.

**Comment by hickst** [comment]
I looked at this again and decided that my previous analysis not very useful: it doesn't relate case of key strings to the count differences. I've re-measured the Uniprot Protein KB because it is the most likely one to have meaningful case differences. Each line of the Uniprot KB has a text key, an ID, and a species:
```
2784260  =  base count                           # count of all entries
2784260  =  base | sort | uniq                   # there are no duplicate entries
2784260  =  base | sort -f | uniq -i             # no duplicates entries even when case is ignored
1834344  =  base | cut f1,2 | sort | uniq        # unique key-ID pairs (ignores species)
1834344  =  base | cut f1,2 | sort -f | uniq -i  # unique key-ID pairs (ignores species). No duplicates even when case is ignored. 
```
But, it seems to me, that what we really want to know about is whether the case of the keys can tell us anything about the species. Again using the Uniprot KB, I created a script to investigate this; which resulted in the following counts by key case category:

Case Category -> Species | Count | % of Total | % of Category
-------------------------- | ------ | ---------- | --------------
UPPERCASE -> Human        |  15668|  0.56| 10.64
Mixed: 1st UPPER -> Human | 173624|  6.24|  9.07
Mixed: 1st lower -> Human |   7684|  0.28|  1.15
lowercase -> Human        |      2|  0.00|  0.00
UPPERCASE -> Other        | 131528|  4.72| 89.36
Mixed: 1st UPPER -> Other |1740309| 62.51| 90.93
Mixed: 1st lower -> Other | 658277| 23.64| 98.85
lowercase -> Other        |  57168|  2.05|100.00
TOTAL                     |2784260

**Comment by MihaiSurdeanu** [comment]
I am not sure this is what we need to know. What we need to know is, if we
preserve casing, how many collisions in grounding ids do we avoid?

**Comment by hickst** [comment]
OK....currently we provide a key "aaa" to the KB and we get back a set of candidate ID-species pairs. So, are you asking how often, in a case-sensitive KB, a lowercased key would return an entirely disjoint set from a key with a different case?

**Comment by hickst** [comment]
So I did the aforementioned analysis of Uniprot case-sensitive keys mapping to entity IDs (see previous comment). I found that 32,843 of our current canonicalized keys would have 2 or 3 different sets of mappings is we used case-sensitive keys. My program does not check whether the different sets are mutually disjoint but, in a random of check of dozens of the sets, I find no sets which overlap. This experiment has convinced me that we can improve our precision by using case sensitive keys for Uniprot proteins.

**Comment by MihaiSurdeanu** [comment]
For this analysis, it would be nice to see 2 numbers:
- Number of keys that would have 2+ sets of mappings is we used
case-sensitive key. Let's call this N. Is this 32,843?
- Number of keys that would have 2+ sets of mappings is we used
*case-insensitive* key. Let's call this M.

If M >> N, then it's worth trying case-sensitive keys.
Mihai

**Comment by hickst** [comment]
> Number of keys that would have 2+ sets of mappings if we used case-sensitive key. Let's call this N. Is this 32,843?

Yes, correct.

> Number of keys that would have 2+ sets of mappings if we used *case-insensitive* key.

I don't understand this question as stated: we **currently** use case insensitive keys and, currently, each key always maps to exactly **one** set of IDs. There are currently 333,303 unique keys in the Uniprot KB (when considered case-insensitively) meaning we have that many distinct mappings of a key to a set of IDs.

IMO, the goal of using case-sensitive keys would be to partition each set of IDs into 2 or more disjoint sets of IDs, thus increasing the precision of the mapping. The investigation above indicates that this partitioning would happen for 32,843 of our current keys. I have not verified that the partitioned sets of IDs would always be disjoint but I have yet to find any partitioned sets (i.e. for a single key) which are not disjoint.

**Comment by hickst** [comment]
Closed by pull request #491
